File: GGAndroid_2.0/GGCore/GGVMCore/src/main/java/io/ganguo/vmodel/ViewModelHelper.java
package io.ganguo.vmodel;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.app.Dialog;
import android.content.Context;
import android.databinding.DataBindingUtil;
import android.databinding.ViewDataBinding;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import io.ganguo.library.BR;
import io.ganguo.library.ui.adapter.v7.hodler.BaseViewHolder;
import io.ganguo.vmodel.adapter.ViewModelAdapter;
import io.ganguo.vmodel.view.AdapterInterface;
import io.ganguo.library.ui.view.DialogInterface;
import io.ganguo.library.ui.view.PopupWindowInterface;
import io.ganguo.library.ui.view.ViewInterface;
import io.ganguo.vmodel.view.ViewInterfaceFactory;

/**
 * ViewModel辅助类，用于ViewModel与View绑定
 * Created by Roger on 6/7/16.
 */
public class ViewModelHelper {

    /**
     * check ViewModel LayoutId
     */
    @SuppressLint("ResourceType")
    private static boolean checkLayoutId(BaseViewModel viewModel) {
        if (viewModel.getItemLayoutId() <= 0) {
            throw new IllegalArgumentException("viewModel not implement getItemLayoutId");
        }
        return true;
    }

    /**
     * 用于与ViewInterface(View,Adapter,Activity)绑定
     */
    @SuppressWarnings("unchecked")
    public static <B extends BaseViewModel> B bind(ViewInterface view, BaseViewModel parent, B viewModel, int variableId) {
        checkLayoutId(viewModel);
        try {
            viewModel.attach(view, variableId);
        } catch (ClassCastException exception) {
            Log.e("bind error:", "viewModel no match with layoutId or error viewInterface");
        }
        // 绑定生命周期
        if (parent != null) {
            bindLifecycle(parent, viewModel);
        }
        return viewModel;
    }

    @SuppressWarnings("unchecked")
    public static <B extends BaseViewModel> B bind(ViewInterface view, B viewModel, int variableId) {
        return bind(view, null, viewModel, variableId);
    }

    public static <B extends BaseViewModel> B bind(ViewInterface view, BaseViewModel parent, B viewModel) {
        return bind(view, parent, viewModel, BR.data);
    }

    public static <B extends BaseViewModel> B bind(ViewInterface view, B viewModel) {
        return bind(view, null, viewModel, BR.data);
    }

    /**
     * 用于与ViewModelAdapter绑定
     *
     * @see AdapterInterface
     */
    public static <B extends BaseViewModel> B bind(ViewModelAdapter adapter, BaseViewHolder vh, BaseViewModel viewModel) {
        checkLayoutId(viewModel);
        return (B) bind(ViewInterfaceFactory.adapterViewFactory(adapter, vh), adapter.getParent(), viewModel, BR.data);
    }


    /**
     * 用于与Dialog绑定
     *
     * @see PopupWindowInterface
     */
    public static <B extends BaseViewModel> B bind(ViewModelPopupWindow popupWindow, B viewModel) {
        return bind(popupWindow, viewModel, BR.data);
    }

    public static <B extends BaseViewModel> B bind(ViewModelPopupWindow popupWindow, B viewModel, int variableId) {
        checkLayoutId(viewModel);
        return bind(ViewInterfaceFactory.popupWindowViewFactory(popupWindow, ViewModelHelper.inflate(popupWindow
                .getContext(), viewModel.getItemLayoutId())), viewModel, variableId);
    }


    /**
     * 用于与Dialog绑定
     *
     * @see DialogInterface
     */
    public static <B extends BaseViewModel> B bind(Dialog dialog, B viewModel) {
        return bind(dialog, viewModel, BR.data);
    }

    public static <B extends BaseViewModel> B bind(Dialog dialog, B viewModel, int variableId) {
        checkLayoutId(viewModel);
        return bind(ViewInterfaceFactory.dialogViewFactory(dialog, ViewModelHelper.inflate(dialog.getContext(), viewModel.getItemLayoutId())), viewModel, variableId);
    }

    /**
     * 用于直接与Binding绑定
     */
    public static <B extends BaseViewModel> B bind(ViewDataBinding binding, BaseViewModel parent, B viewModel, int variableId) {
        checkLayoutId(viewModel);
        return bind(ViewInterfaceFactory.viewFactory(binding), parent, viewModel, variableId);
    }

    public static <B extends BaseViewModel> B bind(ViewDataBinding binding, B viewModel, int variableId) {
        checkLayoutId(viewModel);
        return bind(ViewInterfaceFactory.viewFactory(binding), null, viewModel, variableId);
    }

    public static <B extends BaseViewModel> B bind(ViewDataBinding binding, BaseViewModel parent, B viewModel) {
        return bind(binding, parent, viewModel, BR.data);
    }

    public static <B extends BaseViewModel> B bind(ViewDataBinding binding, B viewModel) {
        return bind(binding, viewModel, BR.data);
    }

    /**
     * 用于create view,并与viewModel绑定
     */
    @Deprecated
    public static <B extends BaseViewModel> B bind(Context context, BaseViewModel parent, B viewModel, int variableId) {
        return bind(ViewInterfaceFactory.viewFactory(context, viewModel.getItemLayoutId()), parent, viewModel, variableId);
    }

    @Deprecated
    public static <B extends BaseViewModel> B bind(Context context, B viewModel, int variableId) {
        return bind(ViewInterfaceFactory.viewFactory(context, viewModel.getItemLayoutId()), viewModel, variableId);
    }

    @Deprecated
    public static <B extends BaseViewModel> B bind(Context context, BaseViewModel parent, B viewModel) {
        return bind(context, parent, viewModel, BR.data);
    }

    @Deprecated
    public static <B extends BaseViewModel> B bind(Context context, B viewModel) {
        return bind(context, viewModel, BR.data);
    }

    @Deprecated
    public static <B extends BaseViewModel> B bind(Activity activity, BaseViewModel parent, B viewModel, int variableId) {
        return bind(ViewInterfaceFactory.viewFactoryForActivity(activity, viewModel.getItemLayoutId()), parent, viewModel, variableId);
    }

    @Deprecated
    public static <B extends BaseViewModel> B bind(Activity activity, B viewModel, int variableId) {
        return bind(ViewInterfaceFactory.viewFactoryForActivity(activity, viewModel.getItemLayoutId()), viewModel, variableId);
    }

    @Deprecated
    public static <B extends BaseViewModel> B bind(Activity activity, BaseViewModel parent, B viewModel) {
        return bind(activity, parent, viewModel, BR.data);
    }

    @Deprecated
    public static <B extends BaseViewModel> B bind(Activity activity, B viewModel) {
        return bind(activity, viewModel, BR.data);
    }


    /**
     * 用于绑定viewModel到view
     */
    public static <B extends BaseViewModel> B bind(View view, BaseViewModel parent, B viewModel, int variableId) {
        ViewDataBinding binding = DataBindingUtil.inflate(LayoutInflater.from(view.getContext()), viewModel.getItemLayoutId(), null, false);
        return bind(ViewInterfaceFactory.viewFactory(binding), parent, viewModel, variableId);
    }


    public static <B extends BaseViewModel> B bind(View view, B viewModel, int variableId) {
        return bind(view, null, viewModel, variableId);
    }

    public static <B extends BaseViewModel> B bind(View view, BaseViewModel parent, B viewModel) {
        ViewDataBinding binding = DataBindingUtil.inflate(LayoutInflater.from(view.getContext()), viewModel.getItemLayoutId(), null, true);
        return bind(ViewInterfaceFactory.viewFactory(binding), parent, viewModel, BR.data);
    }

    public static <B extends BaseViewModel> B bind(View View, B viewModel) {
        return bind(View, null, viewModel);
    }


    /**
     * 用于绑定viewModel到view group container
     */
    public static <B extends BaseViewModel> B bind(ViewGroup container, BaseViewModel parent, B viewModel, int variableId) {
        ViewDataBinding binding = DataBindingUtil.inflate(LayoutInflater.from(container.getContext()), viewModel.getItemLayoutId(), null, false);
        container.addView(binding.getRoot());
        return bind(ViewInterfaceFactory.viewFactory(binding), parent, viewModel, variableId);
    }

    public static <B extends BaseViewModel> B bind(ViewGroup container, B viewModel, int variableId) {
        return bind(container, null, viewModel, variableId);
    }

    public static <B extends BaseViewModel> B bind(ViewGroup container, BaseViewModel parent, B viewModel) {
        ViewDataBinding binding = DataBindingUtil.inflate(LayoutInflater.from(container.getContext()), viewModel.getItemLayoutId(), container, true);
        return bind(ViewInterfaceFactory.viewFactory(binding), parent, viewModel, BR.data);
    }

    public static <B extends BaseViewModel> B bind(ViewGroup container, B viewModel) {
        return bind(container, null, viewModel);
    }

    public static ViewDataBinding inflate(Context context, int LayoutId) {
        return DataBindingUtil.inflate(LayoutInflater.from(context), LayoutId, null, false);
    }

    /**
     * 将子ViewModel绑定父ViewModel生命周期
     */
    @SuppressWarnings("unchecked")
    public static void bindLifecycle(BaseViewModel parent, final BaseViewModel child) {
        if (parent != null && child != null) {
            child.getLifecycleHelper().bindParent(parent);
        } else {
            throw new NullPointerException((parent == null ? "parent" : "child") + "is null");
        }
    }

}

File: GGAndroid_3.0/core/frame/mvvm-v1/src/main/java/io/ganguo/viewmodel/core/ViewModelHelper.kt
package io.ganguo.viewmodel.core

import android.annotation.SuppressLint
import android.app.Dialog
import android.content.Context
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import io.ganguo.adapter.BR
import io.ganguo.adapter.hodler.ItemViewHolder
import io.ganguo.viewmodel.core.adapter.ViewModelAdapter
import io.ganguo.viewmodel.core.creator.ViewInterfaceFactory
import io.ganguo.viewmodel.core.viewinterface.ViewInterface

/**
 * <pre>
 *     @author : leo
 *     time   : 2020/05/29
 *     desc   : ViewModel辅助类，用于ViewModel与View绑定
 * </pre>
 */
object ViewModelHelper {


    /**
     * check ViewModel LayoutId
     * @param viewModel BaseViewModel<*>?
     * @return Boolean
     */
    @SuppressLint("ResourceType")
    private fun checkLayoutId(viewModel: BaseViewModel<*>?): Boolean {
        require(viewModel!!.layoutId > 0) { "viewModel not implement getItemLayoutId" }
        return true
    }

    /**
     * 用于与ViewInterface(View,Adapter,Activity)绑定
     * @param view ViewInterface<*>
     * @param parent BaseViewModel<*>?
     * @param viewModel B
     * @param variableId Int
     * @return B
     */
    fun <B : BaseViewModel<*>> bind(view: ViewInterface<*>, parent: BaseViewModel<*>?, viewModel: B, variableId: Int): B {
        checkLayoutId(viewModel)
        try {
            viewModel.attach(view, variableId)
        } catch (exception: ClassCastException) {
            Log.e("bind error:", "viewModel no match with layoutId or error viewInterface")
        }
        // 绑定生命周期
        parent?.let { bindLifecycle(it, viewModel) }
        return viewModel
    }

    fun <B : BaseViewModel<*>> bind(view: ViewInterface<*>, viewModel: B, variableId: Int): B {
        return bind(view, null, viewModel, variableId)
    }

    fun <B : BaseViewModel<*>> bind(view: ViewInterface<*>, parent: BaseViewModel<*>?, viewModel: B): B {
        return bind(view, parent, viewModel, BR.data)
    }

    fun <B : BaseViewModel<*>> bind(view: ViewInterface<*>, viewModel: B): B {
        return bind(view, null, viewModel, BR.data)
    }


    fun <V : ViewInterface<*>> bind(adapter: ViewModelAdapter, vh: ItemViewHolder<ViewDataBinding>, viewModel: BaseViewModel<V>): BaseViewModel<V> {
        return bind(ViewInterfaceFactory.adapterViewFactory(adapter, vh), adapter.parent, viewModel, BR.data)
    }

    /**
     * 绑定ViewModel到Dialog
     * @param dialog Dialog
     * @param viewModel B
     * @return B
     */
    fun <B : BaseViewModel<*>> bind(dialog: Dialog, viewModel: B): B {
        return bind(dialog, viewModel, BR.data)
    }

    fun <B : BaseViewModel<*>> bind(dialog: Dialog, viewModel: B, variableId: Int): B {
        checkLayoutId(viewModel)
        return bind(ViewInterfaceFactory.dialogViewFactory(dialog, inflate(dialog.context, viewModel.layoutId)), viewModel, variableId)
    }


    /**
     * 绑定ViewModel到ViewDataBinding
     * @param binding ViewDataBinding
     * @param parent BaseViewModel<*>?
     * @param viewModel B
     * @param variableId Int
     * @return B
     */
    fun <B : BaseViewModel<*>> bind(binding: ViewDataBinding, parent: BaseViewModel<*>?, viewModel: B, variableId: Int): B {
        checkLayoutId(viewModel)
        return bind(ViewInterfaceFactory.viewFactory(binding), parent, viewModel, variableId)
    }

    fun <B : BaseViewModel<*>> bind(binding: ViewDataBinding, viewModel: B, variableId: Int): B {
        checkLayoutId(viewModel)
        return bind(ViewInterfaceFactory.viewFactory(binding), null, viewModel, variableId)
    }

    fun <B : BaseViewModel<*>> bind(binding: ViewDataBinding, parent: BaseViewModel<*>?, viewModel: B): B {
        return bind(binding, parent, viewModel, BR.data)
    }

    fun <B : BaseViewModel<*>> bind(binding: ViewDataBinding, viewModel: B): B {
        return bind(binding, viewModel, BR.data)
    }

    /**
     * 绑定viewModel到view
     * @param view View
     * @param parent BaseViewModel<*>?
     * @param viewModel B
     * @param variableId Int
     * @return B
     */
    fun <B : BaseViewModel<*>> bind(view: View, parent: BaseViewModel<*>?, viewModel: B, variableId: Int): B {
        val binding = DataBindingUtil.inflate<ViewDataBinding>(LayoutInflater.from(view.context), viewModel.layoutId, null, false)
        return bind(ViewInterfaceFactory.viewFactory(binding), parent, viewModel, variableId)
    }


    fun <B : BaseViewModel<*>> bind(view: View, viewModel: B, variableId: Int): B {
        return bind(view, null, viewModel, variableId)
    }

    fun <B : BaseViewModel<*>> bind(view: View, parent: BaseViewModel<*>?, viewModel: B): B {
        val binding = DataBindingUtil.inflate<ViewDataBinding>(LayoutInflater.from(view.context), viewModel.layoutId, null, true)
        return bind(ViewInterfaceFactory.viewFactory(binding), parent, viewModel, BR.data)
    }

    fun <B : BaseViewModel<*>> bind(view: View, viewModel: B): B {
        return bind(view, null, viewModel)
    }

    /**
     * 绑定到ViewModel到ViewGroup
     * @param container ViewGroup
     * @param viewModel B
     * @return B
     */
    fun <B : BaseViewModel<*>> bind(container: ViewGroup, parent: BaseViewModel<*>?, viewModel: B, variableId: Int): B {
        val binding = DataBindingUtil.inflate<ViewDataBinding>(LayoutInflater.from(container.context), viewModel.layoutId, null, false)
        container.addView(binding.root)
        return bind(ViewInterfaceFactory.viewFactory(binding), parent, viewModel, variableId)
    }


    fun <B : BaseViewModel<*>> bind(container: ViewGroup, viewModel: B, variableId: Int): B {
        return bind(container, null, viewModel, variableId)
    }

    fun <B : BaseViewModel<*>> bind(container: ViewGroup, parent: BaseViewModel<*>?, viewModel: B): B {
        val binding = DataBindingUtil.inflate<ViewDataBinding>(LayoutInflater.from(container.context), viewModel.layoutId, container, true)
        return bind(ViewInterfaceFactory.viewFactory(binding), parent, viewModel, BR.data)
    }

    fun <B : BaseViewModel<*>> bind(container: ViewGroup, viewModel: B): B {
        return bind(container, null, viewModel)
    }

    /**
     * 将子ViewModel绑定父ViewModel生命周期
     * @param parent BaseViewModel<*>?
     * @param child BaseViewModel<*>?
     */
    private fun bindLifecycle(parent: BaseViewModel<*>?, child: BaseViewModel<*>?) {
        if (parent != null && child != null) {
            child.bindLifecycle(parent)
        } else {
            throw NullPointerException((if (parent == null) "parent" else "child") + "is null")
        }
    }

    /**
     * 绑定单个ViewModel
     * @param context Context
     * @param parent BaseViewModel<*>?
     * @param viewModel B
     * @return BaseViewModel<*>
     */
    fun <B : BaseViewModel<*>> bind(context: Context, parent: BaseViewModel<*>?, viewModel: B): BaseViewModel<*> {
        val viewDataBinding = inflate(context, viewModel.layoutId)
        return bind(viewDataBinding, parent, viewModel)
    }


    /**
     * 创建ViewDataBinding
     * @param context Context
     * @param layoutId Int
     * @return ViewDataBinding
     */
    fun inflate(context: Context, layoutId: Int): ViewDataBinding {
        return DataBindingUtil.inflate(LayoutInflater.from(context), layoutId, null, false)
    }

}